### ISSUE TITLE
Memory leak in case of high load

### ISSUE THREAD

**Issue by tillschaefer** [deepch/RTSPtoWeb#415]
Running RTSPtoWeb with many camera streams (above 25 here) causes the memory to rise until OOM occurs. 

I have diagnosed the problem and found the trigger of the memory leak in the function `func (client *RTSPClient) startStream()` of vdk. If the OutgoingProxyQueue of the RTSPClient is full, the the succeeding return causes the stream to be restarted since SignalStreamRTPStop is emitted. However, It seems that not all resources are freed correctly. 

To confirm the analysis, I have replaced the return in the case of a full channel with a continue. In this case no leak occurs and the memory consumption of the server is stable. 

Relevant part of the code (vdk, format/rtspv2/client.go):
        
    if client.options.OutgoingProxy {
      if len(client.OutgoingProxyQueue) < 2000 {
        client.OutgoingProxyQueue <- &content
      } else {
        client.Println("RTSP Client OutgoingProxy Chanel Full")
        return    
      }
    }
    pkt, got := client.RTPDemuxer(&content)

**Comment by tillschaefer** [comment]
Digging deeper with pprof the majority of the memory is allocated in vdks format/rtspv2/client.go: 

           .          .    710:                                   if client.fuStarted {
           .     6.84MB    711:                                           client.BufferRtpPacket.Write(content[offset+2 : end])
           .          .    712:                                           if isEnd {
           .          .    713:                                                   client.fuStarted = false
           .          .    714:                                                   naluTypef := client.BufferRtpPacket.Bytes()[0] & 0x1f
           .          .    715:                                                   if naluTypef == 7 || naluTypef == 9 {
           .          .    716:                                                           bufered, _ := h264parser.SplitNALUs(append([]byte{0, 0, 0, 1}, client.BufferRtpPacket.Bytes()...))
           .          .    717:                                                           for _, v := range bufered {
           .          .    718:                                                                   naluTypefs := v[0] & 0x1f
           .          .    719:                                                                   switch {
           .          .    720:                                                                   case naluTypefs == 5:
           .          .    721:                                                                           client.BufferRtpPacket.Reset()
           .          .    722:                                                                           client.BufferRtpPacket.Write(v)
           .          .    723:                                                                           naluTypef = 5
           .          .    724:                                                                   case naluTypefs == 7:
           .          .    725:                                                                           client.CodecUpdateSPS(v)
           .          .    726:                                                                   case naluTypefs == 8:
           .          .    727:                                                                           client.CodecUpdatePPS(v)
           .          .    728:                                                                   }
           .          .    729:                                                           }
           .          .    730:                                                   }
         2MB        2MB    731:                                                   retmap = append(retmap, &av.Packet{
      1.22GB     1.22GB    732:                                                           Data:            append(binSize(client.BufferRtpPacket.Len()), client.BufferRtpPacket.Bytes()...),
           .          .    733:                                                           CompositionTime: time.Duration(1) * time.Millisecond,
           .          .    734:                                                           Duration:        time.Duration(float32(timestamp-client.PreVideoTS)/90) * time.Millisecond,
           .          .    735:                                                           Idx:             client.videoIDX,
           .          .    736:                                                           IsKeyFrame:      naluTypef == 5,
           .          .    737:                                                           Time:            time.Duration(timestamp/90) * time.Millisecond,

**Comment by deepch** [comment]
share you streams and config Im fix it

**Comment by deepch** [comment]
check lock loop add continue after https://github.com/deepch/RTSPtoWeb/blob/c15b3b04ee9fc2e4cd43aec4031a5296cfb73cf7/streamCore.go#L140C3-L140C53

**Comment by deepch** [comment]
please get report if usage 8G+

**Comment by tillschaefer** [comment]
> check lock loop add continue after https://github.com/deepch/RTSPtoWeb/blob/c15b3b04ee9fc2e4cd43aec4031a5296cfb73cf7/streamCore.go#L140C3-L140C53

If i skip the packetAv Processing (i.e. discard the package, i got meny ErrorStreamNoVideo logs), the memory stays way below 100 MiB. 
Does this mean this might be only a buffer bloat caused by some performance bottleneck while sending the packages?

The interesting part is, that I can avoid the issue by starting multiple instances of the the server in parallel. Thus, we might se some lock contention here. For example, the function StreamChannelCast is locked on the StorageST struct, i.e. we have effectively a global lock at this place.

**Comment by tillschaefer** [comment]
So i used a profiler to see where all the time is spend. Most of the time is caused by muxing the HLS Stream M3u8. This even happens despite no HLS Stream is actually streamed. Only WebRtc streams are actually connected. As far as I have seen the Streams dont even have an audio channel at all. Thus, it might benefit if we only do the HLS / Audio Muxing when it is actually needed. 

    (pprof) top20
    Showing nodes accounting for 31.30s, 62.26% of 50.27s total
    Dropped 515 nodes (cum <= 0.25s)
    Showing top 20 nodes out of 162
          flat  flat%   sum%        cum   cum%
         9.51s 18.92% 18.92%      9.51s 18.92%  runtime.memmove
         4.71s  9.37% 28.29%      4.77s  9.49%  runtime.cgocall
         2.56s  5.09% 33.38%      2.93s  5.83%  runtime.findObject
         2.03s  4.04% 37.42%      8.43s 16.77%  runtime.scanobject
         1.50s  2.98% 40.40%      1.50s  2.98%  runtime.stdcall2
         1.14s  2.27% 42.67%      2.27s  4.52%  runtime.greyobject
         1.13s  2.25% 44.92%     14.93s 29.70%  runtime.concatstrings
         0.99s  1.97% 46.89%      0.99s  1.97%  github.com/pion/srtp/v2.xorBytes
         0.95s  1.89% 48.78%     20.06s 39.90%  main.(*MuxerHLS).UpdateIndexM3u8
         0.89s  1.77% 50.55%      0.89s  1.77%  strconv.rightShift
         0.71s  1.41% 51.96%      6.50s 12.93%  runtime.mallocgc
         0.68s  1.35% 53.31%      0.68s  1.35%  runtime.markBits.isMarked (inline)
         0.65s  1.29% 54.61%      0.65s  1.29%  runtime._ExternalCode
         0.64s  1.27% 55.88%      2.30s  4.58%  runtime.(*mheap).allocSpan
         0.64s  1.27% 57.15%      0.64s  1.27%  runtime.heapBits.bits (inline)
         0.55s  1.09% 58.25%      0.55s  1.09%  crypto/sha1.blockAVX2
         0.55s  1.09% 59.34%      0.58s  1.15%  strconv.(*decimal).Assign
         0.50s  0.99% 60.33%      0.50s  0.99%  runtime.stdcall6
         0.49s  0.97% 61.31%      2.20s  4.38%  runtime.sweepone
         0.48s  0.95% 62.26%      0.48s  0.95%  crypto/aes.encryptBlockAsm

![image](https://github.com/deepch/RTSPtoWeb/assets/3009428/921c2424-3648-4a7d-ad60-621b999f9643)

**Comment by deepch** [comment]
try comment line //Storage.StreamChannelCast(streamID, channelID, packetAV) and test again

**Comment by deepch** [comment]
Yes, it can be done better, but for hls you need at least 4 segments of 4 seconds to render the video.

**Comment by tillschaefer** [comment]
> try comment line //Storage.StreamChannelCast(streamID, channelID, packetAV) and test again

I think the issue here is quite the opposite. If I just keep the above line (an the keyTest timer) everything runs smooth, otherwise the HLS Muxer is always run (even without any connected client) and that causes the Buffer / OutgoingPacketQueue channel to blow up. Is there a reason the hls clients are not registered the same way as MSE and WebRtc clients via Storage.ClientAdd? This would allow the execution of the HLS Muxer only in cases the HLS stream is actually requested.

**Comment by tillschaefer** [comment]
Ok, I have been investigating the problem a bit more and there is indeed a memory leak, which is more likely unter high load (disconnecting Clients):

The gin context in
[func HTTPAPIServerStreamWebRTC(c *gin.Context) {](https://github.com/deepch/RTSPtoWeb/blob/93ae33bbe4a0a75a3161ad22e475421ff93b376a/apiHTTPWebRTC.go#L12)
is only valid until the function returns. However the goroutine is still alive afterwards and the line https://github.com/deepch/RTSPtoWeb/blob/93ae33bbe4a0a75a3161ad22e475421ff93b376a/apiHTTPWebRTC.go#L70C1-L70C1  does not reliably delete the clients, because in some cases the streamID and clientID are sometimes overwritten with different values.

**Comment by GyeongHoKim** [comment]
related to #496
